### ISSUE TITLE
Reverse phase-encode fmap file not being handled

### ISSUE THREAD

**Issue by dkp** [BIDS-Apps/MRtrix3_connectome#37]
```
9659:feckless:working/trix3> docker run -i --rm -v /Volumes/Main/Working/trix3:/bids_dataset -v /Volumes/Main/Working/trix3/derivatives:/outputs bids/mrtrix3_connectome /bids_dataset /output participant --participant_label 330 --parcellation desikan --debug
mrtrix3_connectome.py:
mrtrix3_connectome.py: Note that this script makes use of commands / algorithms that have relevant articles for citation; INCLUDING FROM EXTERNAL SOFTWARE PACKAGES. Please consult the help page (-help option) for more information.
mrtrix3_connectome.py:
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:66): bids-validator -> bids-validator
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from run.py:375): Command bids-validator not found in MRtrix3 bin/ directory
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:377): bids-validator -> bids-validator
mrtrix3_connectome.py: [DEBUG] run._shebang() (from run.py:67): File "bids-validator": string "#!/usr/bin/env node": ['/usr/bin/env', 'node']
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:1018): To execute: [['/usr/bin/env', 'node', '/usr/local/bin/bids-validator', '/bids_dataset']]
Command:  bids-validator /bids_dataset
mrtrix3_connectome.py: Commencing execution for subject sub-330
mrtrix3_connectome.py: N4BiasFieldCorrection and ROBEX found; will use for bias field correction and brain extraction
mrtrix3_connectome.py: [DEBUG] (from mrtrix3_connectome.py:128) 'reconall_multithread_options' =  -parallel
mrtrix3_connectome.py: Generated temporary directory: /output
mrtrix3_connectome.py: Importing DWI data into temporary directory
mrtrix3_connectome.py: [DEBUG] path.toTemp() (from mrtrix3_connectome.py:211): dwi1.mif -> /output/dwi1.mif
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:351): mrconvert -> mrconvert
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from run.py:57): Version-matched executable for mrconvert: /mrtrix3/bin/mrconvert
mrtrix3_connectome.py: [DEBUG] run._shebang() (from run.py:67): File "/mrtrix3/bin/mrconvert": Not a text file
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:211): To execute: [['/mrtrix3/bin/mrconvert', '/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz', '-fslgrad', '/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.bvec', '/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.bval', '-json_import', '/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.json', '/output/dwi1.mif', '-info']]
Command:  mrconvert /bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz -fslgrad /bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.bvec /bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.bval -json_import /bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.json /output/dwi1.mif
          mrconvert: [INFO] opening image "/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz"...
          mrconvert: [INFO] Axes and transform of image "/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz" altered to approximate RAS coordinate system
          mrconvert: [INFO] image "/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz" opened with dimensions 128x128x74x32, voxel spacing 2x2x2x11.5, datatype Int16LE
          mrconvert: [INFO] found 32x4 diffusion gradient table
          mrconvert: [100%] uncompressing image "/bids_dataset/sub-330/dwi/sub-330_acq-AP_dwi.nii.gz"
          mrconvert: [INFO] creating image "/output/dwi1.mif"...
          mrconvert: [INFO] image "/output/dwi1.mif" created with dimensions 128x128x74x32, voxel spacing 2x2x2x11.5, datatype Int16LE
          mrconvert: [100%] copying from "/bids_data...i/sub-330_acq-AP_dwi.nii.gz" to "/output/dwi1.mif"
mrtrix3_connectome.py: Importing fmap data into temporary directory
mrtrix3_connectome.py: [DEBUG] path.newTemporary() (from image.py:11): /output/mrtrix-tmp-Y3VRMU.json
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:351): mrinfo -> mrinfo
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from image.py:12): Version-matched executable for mrinfo: /mrtrix3/bin/mrinfo
mrtrix3_connectome.py: [DEBUG] run.exeName() (from image.py:12): /mrtrix3/bin/mrinfo -> /mrtrix3/bin/mrinfo
mrtrix3_connectome.py: Loading header for image file '/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz'
mrtrix3_connectome.py: [DEBUG] image.__init__() (from mrtrix3_connectome.py:236): ['/mrtrix3/bin/mrinfo', '/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz', '-json_all', '/output/mrtrix-tmp-Y3VRMU.json']
mrtrix3_connectome.py: [DEBUG] image.__init__() (from mrtrix3_connectome.py:236): {'_intensity_offset': 0.0, '_format': u'NIfTI-1.1 (GZip compressed)', '_transform': [[1.0, 0.0, -0.0, -125.80451965332], [0.0, 1.0, -0.0, -133.989959716797], [-0.0, 0.0, 1.0, -61.0282020568848], [0.0, 0.0, 0.0, 1.0]], '_intensity_scale': 1.0, '_datatype': u'Int16LE', '_name': u'/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz', '_keyval': {u'comments': u'TE=86;Time=143557.320;phase=0;dwell=0.470'}, '_size': [128, 128, 74, 2], '_spacing': [2.0, 2.0, 2.0, 11.5], '_strides': [-1, 2, 3, 4]}
mrtrix3_connectome.py: [DEBUG] path.toTemp() (from mrtrix3_connectome.py:242): fmap1.mif -> /output/fmap1.mif
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:351): mrconvert -> mrconvert
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from run.py:57): Version-matched executable for mrconvert: /mrtrix3/bin/mrconvert
mrtrix3_connectome.py: [DEBUG] run._shebang() (from run.py:67): File "/mrtrix3/bin/mrconvert": Not a text file
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:242): To execute: [['/mrtrix3/bin/mrconvert', '/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz', '-json_import', '/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.json', '-set_property', 'dw_scheme', '0,0,1,0\\n0,0,1,0', '/output/fmap1.mif', '-info']]
Command:  mrconvert /bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz -json_import /bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.json -set_property dw_scheme "0,0,1,0\n0,0,1,0" /output/fmap1.mif
          mrconvert: [INFO] opening image "/bids_dataset/sub-330/fmap/**sub-330_dir-PA_epi.nii.gz**"...
          mrconvert: [INFO] Axes and transform of image "/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz" altered to approximate RAS coordinate system
          mrconvert: [INFO] image "/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz" opened with dimensions 128x128x74x2, voxel spacing 2x2x2x11.5, datatype Int16LE
          mrconvert: [100%] uncompressing image "/bids_dataset/sub-330/fmap/sub-330_dir-PA_epi.nii.gz"
          mrconvert: [INFO] creating image "/output/fmap1.mif"...
          mrconvert: [INFO] image "/output/fmap1.mif" created with dimensions 128x128x74x2, voxel spacing 2x2x2x11.5, datatype Int16LE
          mrconvert: [100%] copying from "/bids_data...p/sub-330_dir-PA_epi.nii.gz" to "/output/fmap1.mif"
mrtrix3_connectome.py: Importing T1 image into temporary directory
mrtrix3_connectome.py: [DEBUG] path.toTemp() (from mrtrix3_connectome.py:256): T1.mif -> /output/T1.mif
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:351): mrconvert -> mrconvert
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from run.py:57): Version-matched executable for mrconvert: /mrtrix3/bin/mrconvert
mrtrix3_connectome.py: [DEBUG] run._shebang() (from run.py:67): File "/mrtrix3/bin/mrconvert": Not a text file
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:256): To execute: [['/mrtrix3/bin/mrconvert', '/bids_dataset/sub-330/anat/sub-330_T1w.nii.gz', '/output/T1.mif', '-info']]
Command:  mrconvert /bids_dataset/sub-330/anat/sub-330_T1w.nii.gz /output/T1.mif
          mrconvert: [INFO] opening image "/bids_dataset/sub-330/anat/sub-330_T1w.nii.gz"...
          mrconvert: [INFO] image "/bids_dataset/sub-330/anat/sub-330_T1w.nii.gz" opened with dimensions 176x240x256, voxel spacing 0.99999350309371948x1.0546875x1.0546875, datatype Int16LE
          mrconvert: [100%] uncompressing image "/bids_dataset/sub-330/anat/sub-330_T1w.nii.gz"
          mrconvert: [INFO] creating image "/output/T1.mif"...
          mrconvert: [INFO] image "/output/T1.mif" created with dimensions 176x240x256, voxel spacing 0.99999350309371948x1.0546875x1.0546875, datatype Int16LE
          mrconvert: [100%] copying from "/bids_data...330/anat/sub-330_T1w.nii.gz" to "/output/T1.mif"
mrtrix3_connectome.py: Changing to temporary directory (/output)
mrtrix3_connectome.py: Concatenating DWI and fmap data for combined pre-processing
mrtrix3_connectome.py: [DEBUG] run.exeName() (from run.py:351): mrcat -> mrcat
mrtrix3_connectome.py: [DEBUG] run.versionMatch() (from run.py:57): Version-matched executable for mrcat: /mrtrix3/bin/mrcat
mrtrix3_connectome.py: [DEBUG] run._shebang() (from run.py:67): File "/mrtrix3/bin/mrcat": Not a text file
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:283): To execute: [['/mrtrix3/bin/mrcat', 'fmap1.mif', 'fmap_cat.mif', '-axis', '3', '-info']]
Command:  mrcat fmap1.mif fmap_cat.mif -axis 3
          mrcat: [ERROR] Expected at least 3 arguments (2 supplied)
mrtrix3_connectome.py:
mrtrix3_connectome.py: [ERROR] Command failed: mrcat fmap1.mif fmap_cat.mif -axis 3 (mrtrix3_connectome.py:283)
mrtrix3_connectome.py: Output of failed command:
                       mrcat: [ERROR] Expected at least 3 arguments (2 supplied)
mrtrix3_connectome.py:
mrtrix3_connectome.py: Changing back to original directory (/)
mrtrix3_connectome.py: Script failed while executing the command: mrcat fmap1.mif fmap_cat.mif -axis 3
mrtrix3_connectome.py: For debugging, inspect contents of temporary directory: /output
```

**Comment by Lestropie** [comment]
Okay, you caught me on this one ><

It looks like you've got only a single EPI volume in the `fmap/` directory (presumably with the opposite phase encoding to the DWIs?). I've made a mistake in the script code by assuming that there will be at least two such volumes, which leads to failure of the *MRtrix3* command `mrcat`.

The *MRtrix3* script `dwipreproc` for which such data are required *is* compatible with such data: it will automatically combine this volume with the *b*=0 volumes from the DWIs to provide to FSL's `topup` tool in estimating the inhomogeneity field. So I will need to update the logic within the script to reflect this possibility.

Thanks for the report!

Rob

**Comment by dkp** [comment]
Many thanks for all your hard work...and your gracious explanations when I was confused.

**Comment by dkp** [comment]
Hi,

I am still running into this problem:
I am using the docker container mrtrix_connectome, pulled today May 5, 2018.
**If I have a reverse phase encode B0 image in the fmap directory, MRtrix fails.**
error.txt attached.
I have removed the reverse phase encode image for testing purposes, but how should I proceed?

Thanks much,

Dianne

[error.txt](https://github.com/BIDS-Apps/MRtrix3_connectome/files/3146236/error.txt)
